onnxruntime/contrib_ops/cpu/reverse.cc

```diff
@@ -0,0 +1,300 @@
+
+ONNX_OPERATOR_KERNEL_EX(
+    Reverse,
+    kMSDomain,
+    1,
+    kCpuExecutionProvider,
+    KernelDefBuilder().TypeConstraint("T", DataTypeImpl::AllTensorTypes()),
+    Reverse);
+
+// some templated aliases
+template <typename T, int rank>
+using EigenTensorMap = Eigen::TensorMap<Eigen::Tensor<T, rank, Eigen::RowMajor, Eigen::DenseIndex>>;
+
+template <typename T, int rank>
+using ConstEigenTensorMap = Eigen::TensorMap<Eigen::Tensor<const T, rank, Eigen::RowMajor, Eigen::DenseIndex>>;
+
+// utility helpers specific to Reverse
+template <int rank>
+Eigen::array<bool, rank> generate_bool_reverse_axes(const std::vector<int64_t>& reverse_axes) {
+  Eigen::array<bool, rank> eigen_reverse_axes;
+
+  // default axes - reverse all axes as per spec
+  if (reverse_axes.size() == 0) {
+    for (int i = 0; i < rank; ++i) {
+      eigen_reverse_axes[i] = true;
+    }
+    return eigen_reverse_axes;
+  }
+
+  // explicit axes given
+  eigen_reverse_axes.fill(false);
+  for (int i = 0; i < reverse_axes.size(); ++i) {
+    const auto& dim = reverse_axes[i];
+    eigen_reverse_axes[dim >= 0 ? dim : dim + rank] = true;
+  }
+  return eigen_reverse_axes;
+}
+
+template <int rank>
+Eigen::DSizes<Eigen::DenseIndex, rank> dims_as_eigen_dsizes(const std::vector<int64_t> dims) {
+  Eigen::DSizes<Eigen::DenseIndex, rank> eigen_dsizes;
+  for (int i = 0; i < rank; ++i) {
+    eigen_dsizes[i] = static_cast<Eigen::DenseIndex>(dims[i]);
+  }
+  return eigen_dsizes;
+}
+
+template <typename T, int rank>
+EigenTensorMap<T, rank> buffer_as_eigen_tensor(T* buffer, const std::vector<int64_t>& dims) {
+  return EigenTensorMap<T, rank>(buffer, dims_as_eigen_dsizes<rank>(dims));
+}
+
+template <typename T, int rank>
+ConstEigenTensorMap<T, rank> buffer_as_const_eigen_tensor(const T* buffer, const std::vector<int64_t>& dims) {
+  return ConstEigenTensorMap<T, rank>(buffer, dims_as_eigen_dsizes<rank>(dims));
+}
+
+Status handle_scalar_tensor(const Tensor* input_tensor, Tensor* output_tensor, const MLDataType& dtype) {
+  // for scalar tensors, "reversing" is just copying the input buffer to the output buffer
+ 
+  // string scalar
+  if (dtype == DataTypeImpl::GetType<std::string>()) {
+    const std::string* src = input_tensor->template Data<std::string>();
+    std::string* dst = output_tensor->template MutableData<std::string>();
+    if (src != dst)
+      std::copy(src, src + 1, dst);
+    return Status::OK();
+  }
+
+  // MLFLoat16 scalar
+  if (dtype == DataTypeImpl::GetType<MLFloat16>()) {
+    const auto* src = input_tensor->Data<MLFloat16>();
+    auto* dst = output_tensor->template MutableData<MLFloat16>();
+    dst[0] = MLFloat16(math::floatToHalf(math::halfToFloat(src[0].val)));
+    return Status::OK();
+  }
+
+  // BFLoat16 scalar
+  if (dtype == DataTypeImpl::GetType<BFloat16>()) {
+    const auto* src = input_tensor->Data<BFloat16>();
+    auto* dst = output_tensor->template MutableData<BFloat16>();
+    dst[0] = BFloat16(src[0].ToFloat());
```

Comment:
>BFloat16(src[0].ToFloat()); [](start = 13, length = 27)

Strange conversion from BFloat16 to float and then back to BFloat16? What is the purpose of this code?

Reply:
This was an over-sight. Fixed, thanks.